### PR TITLE
feat(jest): support artifact collection for multiple retries with jest.retryTimes(n)

### PR DESCRIPTION
**Pull request by noomorph**
- [x] This change has been discussed in issue #2073 and the solution has been agreed upon with maintainers.

---

**Description:**

Adds support `jest.retryTimes(n)` when used with `jest-circus`.

For example, if a `Flaky test` fails 2 times and then passes on the third, the artifacts will be saved like this:

```
artifacts/<auto-generated-dir>/✗ Flaky test/...
artifacts/<auto-generated-dir>/✗ Flaky test (2)/...
artifacts/<auto-generated-dir>/✓ Flaky test (3)/...
```

This pull request has unit and E2E tests for the feature.

### CODE REVIEW

File: detox/src/artifacts/utils/ArtifactPathBuilder.test.js [hunk 1]
```diff
@@ -73,6 +73,20 @@ describe(ArtifactPathBuilder, () => {
...
+
+      expect(asPosixPath(artifactPath)).toBe('/tmp/subdir/✗ test (2)/1.log');
+    });
+
+    it('should append (3) invocations count for a test that passes for the third time', () => {
+      const testSummary = {title: '', fullName: 'test', status: 'passed', invocations: 3 };
+      const artifactPath = pathBuilder.buildPathForTestArtifact('1.log', testSummary);
+
+      expect(asPosixPath(artifactPath)).toBe('/tmp/subdir/✓ test (3)/1.log');
+    });
```
Comment (d4vidi) [inline detox/src/artifacts/utils/ArtifactPathBuilder.test.js:89]: Just wondering: isn't this test redundant?...

File: detox/test/e2e/27.retries.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  });
+
+  it('should fail twice and pass once', async () => {
+    const matcher = --counter > 0
+      ? by.text('Not existing')
+      : by.text('Sanity');
+
+    await element(matcher).tap();
+  });
+});
```
Comment (d4vidi) [inline detox/test/e2e/27.retries.js:17]: Nice, but here too (much like with screenshot tests) we lack the testing of the final artifacts...

File: detox/test/e2e/config.js [hunk 3]
```diff
@@ -2,7 +2,7 @@ module.exports = {
   "rootDir": "../..",
   "testEnvironment": "./test/e2e/environment.js",
   "testRunner": "./test/node_modules/jest-circus/runner",
-  "setupFilesAfterEnv": ['./test/e2e/init-coverage.js'],
+  "setupFilesAfterEnv": [],
```
Comment (d4vidi) [inline detox/test/e2e/config.js:5]: Not really related, right?

File: detox/test/e2e/27.retries.js (hunk 2 above)
Comment (d4vidi) [reply detox/test/e2e/27.retries.js:17]: Also, I didn't see anything related in the build's artifacts. Have I been looking in the wrong place?

File: detox/src/artifacts/utils/ArtifactPathBuilder.test.js (hunk 1 above)
Comment (noomorph) [reply detox/src/artifacts/utils/ArtifactPathBuilder.test.js:89]: I adhere to the blackbox approach, @d4vidi.
In my opinion, there were two ways to implement this.
Either we omit ` (${invocations})` suffix, when the status is `passed` or we add it nevertheless.
Therefore, I **fixate** the behavior and confirm that I as an author intend it to be this way.

File: detox/test/e2e/config.js (hunk 3 above)
Comment (noomorph) [reply detox/test/e2e/config.js:5]: Yeah, but I've noticed that this file is redundant already. It should have gone with my other Coveralls fixes, but I missed it.

File: detox/test/e2e/27.retries.js (hunk 2 above)
Comment (noomorph) [reply detox/test/e2e/27.retries.js:17]: Hmmm... well, I can add checks, theoretically.

File: detox/src/artifacts/utils/ArtifactPathBuilder.test.js (hunk 1 above)
Comment (d4vidi) [reply detox/src/artifacts/utils/ArtifactPathBuilder.test.js:89]: Actually, I see now that I missed the difference in `status` value. You can ignore my comment.

File: detox/test/e2e/27.retries.js (hunk 2 above)
Comment (d4vidi) [reply detox/test/e2e/27.retries.js:17]: So action items here:
1. Rename to `27.retries.test.js`
2. Add artifacts validation?

Comment (noomorph) [reply detox/test/e2e/27.retries.js:17]: Thanks for pointing out, fixing.

**Pull request merged**